### ISSUE TITLE
Flexible downtime is expired at end_time, not trigger_time+duration

### ISSUE THREAD

**Issue by Mordecaine** [Icinga/icinga2#5277]
When I create a flexible downtime, I define a timerange and a duration.
When the service runs in the status Not-OK between the timerange, the icinga should set the server to maintanance mode until the duration time defines.
At moment, the icinga set the server to maintanance mode until the timerange ends and ignores the duration.

API Command POST:
```
curl -k -s -u api_user:******** -H 'Accept: application/json' -X POST 'https://monitoring-slave.example.net:5665/v1/actions/schedule-downtime?type=Host&filter=host.name==%22testserver.example.net' -d '{ "start_time": 1495440600 , "end_time": 1495441200 , "fixed": false, "duration": 1800, "author": "api_user", "comment": "Instance stopped" }' | python -m json.tool
{
    "results": [
        {
            "code": 200.0,
            "legacy_id": 97.0,
            "name": "testserver.example.net!monitoring-slave-1495440064-95",
            "status": "Successfully scheduled downtime 'testserver.example.net!onitorin!monitoring-slave-1495440064-95' for object 'testserver.example.net'."
        }
    ]
}
```

API Command GET:
```
{
    "results": [
        {
            "attrs": {
                "__name": "testserver.example.net!monitoring-slave-1495440064-95",
                "active": true,
                "author": "agile",
                "comment": "Instance stopped from Agility",
                "config_owner": "",
                "duration": 1800.0,
                "end_time": 1495441200.0,
                "entry_time": 1495440064.508164,
                "fixed": false,
                "ha_mode": 0.0,
                "host_name": "testserver.example.net",
                "legacy_id": 97.0,
                "name": "monitoring-slave-1495440064-95",
                "original_attributes": null,
                "package": "_api",
                "paused": false,
                "scheduled_by": "",
                "service_name": "",
                "start_time": 1495440600.0,
                "templates": [
                    "monitoring-slave-1495440064-95"
                ],
                "trigger_time": 0.0,
                "triggered_by": "",
                "triggers": [],
                "type": "Downtime",
                "version": 1495440064.508185,
                "was_cancelled": false,
                "zone": "community"
            },
            "joins": {},
            "meta": {},
            "name": "testserver.example.net!monitoring-slave-1495440064-95",
            "type": "Downtime"
        }
    ]
}
```

## Expected Behavior
The Icinga should the set the server in maintanance mode until the duration time ends.

## Steps to Reproduce (for bugs)
1. define a flexible downtime for a host (for example 2pm to 2:10pm and a duration of 30 minutes)
2. disable ICMP echo requests on the host beween 2pm and 2:10pm
3. The server will run into a downtime. The downtime will end 2:10pm and not 30 minutes after the check was Not-OK

## Your Environment
#icinga2 - The Icinga 2 network monitoring daemon (version: v2.6.2)

Copyright (c) 2012-2017 Icinga Development Team (https://www.icinga.com/)
License GPLv2+: GNU GPL version 2 or later <http://gnu.org/licenses/gpl2.html>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.

Application information:
  Installation root: /usr
  Sysconf directory: /etc
  Run directory: /run
  Local state directory: /var
  Package data directory: /usr/share/icinga2
  State path: /var/lib/icinga2/icinga2.state
  Modified attributes path: /var/lib/icinga2/modified-attributes.conf
  Objects path: /var/cache/icinga2/icinga2.debug
  Vars path: /var/cache/icinga2/icinga2.vars
  PID path: /run/icinga2/icinga2.pid

System information:
  Platform: Red Hat Enterprise Linux Server
  Platform version: 7.3 (Maipo)
  Kernel: Linux
  Kernel version: 3.10.0-514.2.2.el7.x86_64
  Architecture: x86_64

Build information:
  Compiler: GNU 4.8.5
  Build host: unknown

**Comment by Raghnall** [comment]
I will like to add that I am also experiencing this same issue on 2.7.1-1.

icinga2 - The Icinga 2 network monitoring daemon (version: r2.7.1-1)

Copyright (c) 2012-2017 Icinga Development Team (https://www.icinga.com/)
License GPLv2+: GNU GPL version 2 or later <http://gnu.org/licenses/gpl2.html>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.

Application information:
  Installation root: /usr
  Sysconf directory: /etc
  Run directory: /run
  Local state directory: /var
  Package data directory: /usr/share/icinga2
  State path: /var/lib/icinga2/icinga2.state
  Modified attributes path: /var/lib/icinga2/modified-attributes.conf
  Objects path: /var/cache/icinga2/icinga2.debug
  Vars path: /var/cache/icinga2/icinga2.vars
  PID path: /run/icinga2/icinga2.pid

System information:
  Platform: Oracle Linux Server
  Platform version: 7.4
  Kernel: Linux
  Kernel version: 3.8.13-118.19.7.el7uek.x86_64
  Architecture: x86_64

Build information:
  Compiler: GNU 4.8.5
  Build host: unknown

**Comment by teclogi** [comment]
Are there any news on this issue?
The problem still persist in version r2.8.1-1 on Debian 9

IcingaWeb2 (version 2.5.1) is displaying then correct end time.

**Comment by terra-nova** [comment]
Hi,

we are also encountering this issue regularly. From glancing over `lib/icinga/downtime.cpp`, it seems that the `Downtime::IsInEffect()` method will always return false if the "trigger window" for a flexible downtime has passed:

```
bool Downtime::IsInEffect() const
{
    double now = Utility::GetTime();

    if (now < GetStartTime() ||
        now > GetEndTime())
        return false;

    if (GetFixed())
        return true;
...
```

I am not sure if I fully understand the logic here, but I think this method should instead look somewhat like this:

```
bool Downtime::IsInEffect() const
{
    double now = Utility::GetTime();

    if (GetFixed()) {a
        return now >= GetStartTime() && now < GetEndTime();
    }

    double triggerTime = GetTriggerTime();

    if (triggerTime == 0)
        return false;

    return (now < triggerTime + GetDuration());
}
```

Regards

[downtime.cpp.issue5722_patch.txt](https://github.com/Icinga/icinga2/files/1889915/downtime.cpp.issue5722_patch.txt)

**Comment by terra-nova** [comment]
@Crunsher Ok, I have created PR #6218.

**Comment by Crunsher** [comment]
Closed via #6218
